presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1411,48 +1414,82 @@ else if (expression instanceof LongLiteral) {
         return orderByExpressions;
     }
 
-    private List<FieldOrExpression> analyzeGroupBy(QuerySpecification node, RelationType tupleDescriptor, AnalysisContext context, List<FieldOrExpression> outputExpressions)
+    private List<List<FieldOrExpression>> analyzeGroupBy(QuerySpecification node, RelationType tupleDescriptor, AnalysisContext context, List<FieldOrExpression> outputExpressions)
     {
-        ImmutableList.Builder<FieldOrExpression> groupByExpressionsBuilder = ImmutableList.builder();
-        if (!node.getGroupBy().isEmpty()) {
-            // Translate group by expressions that reference ordinals
-            for (Expression expression : node.getGroupBy()) {
-                // first, see if this is an ordinal
-                FieldOrExpression groupByExpression;
+        if (node.getGroupBy().isEmpty()) {
+            analysis.setGroupingSets(node, ImmutableList.of());
+            return ImmutableList.of();
+        }
 
-                if (expression instanceof LongLiteral) {
-                    long ordinal = ((LongLiteral) expression).getValue();
-                    if (ordinal < 1 || ordinal > outputExpressions.size()) {
-                        throw new SemanticException(INVALID_ORDINAL, expression, "GROUP BY position %s is not in select list", ordinal);
-                    }
+        ImmutableList.Builder<List<FieldOrExpression>> groupingSetsBuilder = ImmutableList.builder();
+        List<GroupingElement> groupingElements = node.getGroupBy().stream()
+                .filter(groupingElement -> !(groupingElement instanceof  SimpleGroupBy))
+                .collect(Collectors.toList());
 
-                    groupByExpression = outputExpressions.get((int) (ordinal - 1));
-                }
-                else {
-                    ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, tupleDescriptor, context);
-                    analysis.addInPredicates(node, expressionAnalysis.getSubqueryInPredicates());
-                    groupByExpression = new FieldOrExpression(expression);
-                }
+        if (groupingElements.size() < node.getGroupBy().size()) {
+            groupingElements.add(new SimpleGroupBy(node.getGroupBy().stream()
```

Comment:
It's not a good idea to mutate a list that results from calling `.collect(Collectors.toList())`. There's no guarantee the list is mutable. From the javadocs of `Collectors.toList()`:

>  Returns a Collector that accumulates the input elements into a
     new List. There are no guarantees on the type, mutability,
     serializability, or thread-safety of the List returned